File: backend_samples/library/javascript/AzureAIVisionFaceDeviceAttestation/src/cert_store.ts
/**
 * Certificate store, keyed by SHA-256 thumbprint of the cert DER. Persistence,
 * key namespace (cert_thumb:*) and TTL are delegated to an injected ClusterStore.
 *
 *   saveCertificate           — creates `cert_thumb:{thumbprint}` on first sight,
 *                               bumps lastVerifiedAt on subsequent calls
 *   getCertificate            — load record by thumbprint (callers compare
 *                               clientId / system fields to detect cross-identity
 *                               reuse)
 *   updateCertificateMetadata — merge fields into .metadata, bump lastVerifiedAt,
 *                               preserve TTL (used to advance lastAssertionSignCount)
 *
 * The cert TTL is bounded by env `CERT_TTL` (default 7 days), so stale records
 * naturally drop and clients are forced through re-registration. The metadata
 * blob holds the platform-specific verdict (PlayIntegrityVerdict /
 * AppAttestVerdict) so post-attestation we retain everything we decoded.
 */
import { computeCertThumbprint } from './cert_utils';
import { trackEvent } from './logging';
import type { ClusterStore, Snapshot } from './store';

/**
 * Certificate data structure for app attestation
 */
export interface CertificateData {
  clientId: string;
  system: 'ios' | 'android';
  thumbprint: string; // SHA256 hex (64 chars)
  publicCert: string; // PEM format
  createdAt: string; // ISO 8601
  lastVerifiedAt: string; // ISO 8601
  metadata?: object;
}

/**
 * Save a certificate through the injected store.
 * @param clientId - Client application ID
 * @param system - Platform ('ios' or 'android')
 * @param publicCert - Certificate in PEM format
 * @param metadata - Optional metadata object
 * @returns Object with thumbprint and isNew flag, or null on error
 */
export async function saveCertificate(
  store: ClusterStore,
  clientId: string,
  system: 'ios' | 'android',
  publicCert: string,
  metadata?: object,
): Promise<{ thumbprint: string; isNew: boolean } | null> {
  // Compute thumbprint (domain concern; the store owns the key schema + TTL).
  const thumbprint = computeCertThumbprint(publicCert);
  if (!thumbprint) {
    trackEvent('CertStore.SaveFail', {
      reason: 'THUMBPRINT_COMPUTE_FAIL',
      system,
    });
    return null;
  }

  // Existing cert: keep the record, just bump lastVerifiedAt. New cert: build a
  // fresh record. Updates preserve expiry; creation uses the default TTL.
  const existing = await store.getCertificate(thumbprint);
  if (existing) {
    if (existing.value.clientId !== clientId || existing.value.system !== system) return null;
    existing.value.lastVerifiedAt = new Date().toISOString();
    const result = await store.updateCertificate(thumbprint, existing.version, existing.value);
    if (result !== 'applied') {
      return null;
    }
    trackEvent('CertStore.Updated', {
      thumbprint,
      clientId,
      system,
      isNew: false,
    });
    return { thumbprint, isNew: false };
  }

  const now = new Date().toISOString();
  const certData: CertificateData = {
    clientId,
    system,
    thumbprint,
    publicCert,
    createdAt: now,
    lastVerifiedAt: now,
    metadata,
  };

  const ok = await store.setCertificate(thumbprint, certData);
  if (!ok) {
    return null;
  }
  trackEvent('CertStore.Created', {
    thumbprint,
    clientId,
    system,
    isNew: true,
  });
  return { thumbprint, isNew: true };
}

/**
 * Get a certificate snapshot from the injected store by thumbprint.
 * @param thumbprint - Certificate thumbprint (SHA-256 hex)
 * @returns Certificate data or null if not found
 */
export async function getCertificate(
  store: ClusterStore,
  thumbprint: string,
): Promise<Snapshot<CertificateData> | null> {
  return store.getCertificate(thumbprint);
}

/**
 * Merge fields into a cert record's `metadata` and bump `lastVerifiedAt`,
 * preserving the existing certificate expiry. Used by the
 * ongoing-assertion path to advance `lastAssertionSignCount` after each
 * successful per-call verification. Returns false on a stale version or a
 * missing/expired record. Storage failures throw; callers must fail closed.
 */
export async function updateCertificateMetadata(
  store: ClusterStore,
  thumbprint: string,
  partialMetadata: Record<string, unknown>,
  snapshot: Snapshot<CertificateData>,
): Promise<boolean> {
  const existing = snapshot.value;

  existing.lastVerifiedAt = new Date().toISOString();
  existing.metadata = {
    ...(existing.metadata as Record<string, unknown> | undefined),
    ...partialMetadata,
  };

  return (await store.updateCertificate(thumbprint, snapshot.version, existing)) === 'applied';
}

File: backend_samples/library/python/AzureAIVisionFaceDeviceAttestation/azure_ai_vision_face_deviceattestation/cert_store.py
"""Certificate records, keyed by the SHA-256 thumbprint of the cert DER.

Domain logic only: the injected :class:`ClusterStore` owns persistence, TTL, and
serialization. A record is a :class:`CertificateData`; re-verifying an existing
cert bumps ``last_verified_at`` (preserving its TTL).
"""

from __future__ import annotations

import logging
import time
from datetime import datetime, timezone
from typing import Any, Optional

from .cert_utils import compute_cert_thumbprint
from .logging import track_event, track_exception
from .store import CertificateData, ClusterStore, Snapshot, StorageError, UpdateResult

log = logging.getLogger("cert_store")


def _iso_now() -> str:
    return (
        datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%S.")
        + f"{int(time.time() * 1000) % 1000:03d}Z"
    )


async def save_certificate(
    store: ClusterStore,
    client_id: str,
    system: str,
    public_cert: str,
    metadata: Optional[dict[str, Any]] = None,
) -> Optional[dict]:
    """Create or bump a certificate record. Returns {thumbprint, isNew} or None."""
    thumbprint = compute_cert_thumbprint(public_cert)
    if not thumbprint:
        log.error("Failed to compute certificate thumbprint")
        track_event(
            "CertStore.SaveFail",
            {"reason": "THUMBPRINT_COMPUTE_FAIL", "clientId": client_id, "system": system},
        )
        return None
    try:
        existing = await store.get_certificate(thumbprint)
        if existing:
            if existing.value.client_id != client_id or existing.value.system != system:
                return None
            existing.value.last_verified_at = _iso_now()
            if await store.update_certificate(thumbprint, existing.version, existing.value) != UpdateResult.APPLIED:
                return None
            track_event(
                "CertStore.Updated",
                {"thumbprint": thumbprint, "clientId": client_id, "system": system, "isNew": False},
            )
            return {"thumbprint": thumbprint, "isNew": False}

        now = _iso_now()
        cert_data = CertificateData(
            client_id=client_id,
            system=system,
            thumbprint=thumbprint,
            public_cert=public_cert,
            created_at=now,
            last_verified_at=now,
            metadata=metadata,
        )
        if not await store.set_certificate(thumbprint, cert_data):
            return None
        track_event(
            "CertStore.Created",
            {"thumbprint": thumbprint, "clientId": client_id, "system": system, "isNew": True},
        )
        return {"thumbprint": thumbprint, "isNew": True}
    except Exception as err:  # noqa: BLE001
        log.error("Error saving certificate: %s", err)
        track_exception(err, {"source": "save_certificate", "thumbprint": thumbprint})
        raise StorageError("Could not save certificate") from err


async def get_certificate(store: ClusterStore, thumbprint: str) -> Optional[Snapshot[CertificateData]]:
    try:
        return await store.get_certificate(thumbprint)
    except Exception as err:  # noqa: BLE001
        log.error("Error retrieving certificate: %s", err)
        track_exception(err, {"source": "get_certificate", "thumbprint": thumbprint})
        raise StorageError("Could not read certificate") from err


async def update_certificate_metadata(
    store: ClusterStore, thumbprint: str, partial_metadata: dict[str, Any], snapshot: Snapshot[CertificateData]
) -> bool:
    """Merge fields into metadata + bump last_verified_at (store preserves TTL)."""
    try:
        existing = snapshot.value
        existing.last_verified_at = _iso_now()
        merged = dict(existing.metadata or {})
        merged.update(partial_metadata)
        existing.metadata = merged
        return await store.update_certificate(thumbprint, snapshot.version, existing) == UpdateResult.APPLIED
    except Exception as err:  # noqa: BLE001
        track_exception(err, {"source": "update_certificate_metadata", "thumbprint": thumbprint})
        raise StorageError("Could not update certificate") from err
